suite/tests/linux/syscall_pwait.c

```diff
@@ -49,40 +49,73 @@
 #include <unistd.h>
 #include <time.h>
 #include <poll.h>
+#include <sched.h>
+#include "thread.h"
+#include "condvar.h"
 
 typedef struct {
     sigset_t *sigmask;
     size_t sizemask;
 } data_t;
 
+typedef struct {
+    pthread_t main_thread;
+    bool delay;
+} args_t;
+
 static struct timespec sleeptime;
+static void *ready_to_listen;
+static args_t a;
 
 static void
 signal_handler(int sig, siginfo_t *siginfo, void *context)
 {
     print("signal received: %d\n", sig);
 }
 
-bool
-kick_off_child_signals()
+THREAD_FUNC_RETURN_TYPE
+kick_off_child_usr1(void *arg)
 {
-    /* waste some time */
-    nanosleep(&sleeptime, NULL);
-
-    int pid = fork();
-    if (pid < 0) {
-        perror("fork error");
-    } else if (pid == 0) {
+    args_t a = *(args_t *)arg;
+    if (a.delay) {
         /* waste some time */
         nanosleep(&sleeptime, NULL);
-        kill(getppid(), SIGUSR2);
+    }
+    pthread_kill(a.main_thread, SIGUSR1);
+    signal_cond_var(ready_to_listen);
+    return THREAD_FUNC_RETURN_ZERO;
+}
+
+THREAD_FUNC_RETURN_TYPE
+kick_off_child_usr2(void *arg)
+{
+    args_t a = *(args_t *)arg;
+    if (a.delay) {
         /* waste some time */
         nanosleep(&sleeptime, NULL);
-        kill(getppid(), SIGUSR1);
-        return true;
     }
+    pthread_kill(a.main_thread, SIGUSR2);
+    signal_cond_var(ready_to_listen);
+    return THREAD_FUNC_RETURN_ZERO;
+}
 
-    return false;
+void
+kick_off_child_signal(int count, pthread_t main_thread, bool delay)
+{
+    reset_cond_var(ready_to_listen);
+    a.main_thread = main_thread;
+    a.delay = delay;
+    if (count == 1) {
+        create_thread(kick_off_child_usr1, &a);
+    } else if (count == 2) {
+        create_thread(kick_off_child_usr2, &a);
+    }
+    if (!delay) {
+        /* This makes sure that the signal is pending
+         * in the kernel after return of this call.
+         */
+        wait_cond_var(ready_to_listen);
+    }
```

Comment:
Missing join_thread(), so there are resource leaks here (zombie threads).  If you want to *not* wait for the child for the delay case maybe you want to create a single, permanent thread up front, and have a shared args struct plus synch to tell it what to do?  Or else join after the syscall.

Reply:
Done.